pkg/provenance/provenance.go

```diff
@@ -147,122 +148,436 @@ func (o ObjectLineage) GetVersions() string {
+		}
+	}
+	return "No version found that matches the query."
 }
-
-//TODO: add optional parameters to spechistory route in apiserver.go, and call this method.
-// Right now it is actually unused
-func (o ObjectLineage) SpecHistoryInterval(vNumStart, vNumEnd int) []string {
-	//order keys so we append in order later, reference: https://blog.golang.org/go-maps-in-action#TOC_7.
-	s := make([]Spec, 0)
-	for _, spec := range o {
-		s = append(s, spec)
-	}
-	sort.Slice(s, func(i, j int) bool {
-		return s[i].Version < s[i].Version
-	})
-	specs := make([]string, 0)
-	for _, spec := range s {
-		if spec.Version >= vNumStart && spec.Version <= vNumEnd {
-			specs = append(specs, spec.String())
+//this is for a field like deployMent where the underyling state or data is a string
+func handleTrivialFields(fieldData, qkey, qval, mkey string) bool{
+	if qkey == mkey && qval == fieldData {
+		return true
+	}
+	return false
+}
+//this is for a field like databases where the underyling state or data is a slice of strings.
+func handleSimpleFields(vStringSlice []string, qkey, qval, mkey string) bool{
+	satisfied := false
+	for _, str := range vStringSlice {
+		if qkey == mkey && qval == str {
+			satisfied = true
+			break
 		}
 	}
-	return specs
+	return satisfied
+}
+func handleCompositeFields(vSliceMap []map[string]string, mapRelationships map[string][][]string, qkey, qval, mkey string) bool{
+	for _, mymap := range vSliceMap { //looping through each elem in the mapSlice
+		// check if there is any
+		// other necessary requirements for this mkey.
+		// if key exists, then multiple attributes have to be satisfied
+		// at once for the query to work.
+
+		//For example, say fields username and password belong to
+		//an attribute in the spec called 'users'. The username and password
+		//must be satisfied together in some element of vSliceMap since.
+		//It cannot be the case where username is found in index 1 of vSliceMapSlice and password cannot
+		//is found in index 2 of vSliceMapSlice.
+
+		// find all the related attributes associated with the top-level specs
+		// attribute mkey. this would be users for the postgres crd example.
+		attributeCombinedQuery, ok := mapRelationships[mkey]
+
+		if ok { //ensure multiple attributes are jointly satisfied
+			jointQueryResults := make([]bool, len(attributeCombinedQuery))
+			//jointQueryResults is a boolean slice that represents the satisfiability
+			//of the joint query. (all need to be true for it to have found qkey to be true)
+			i := 0
+			for _, pair := range attributeCombinedQuery {
+				qckey := pair[0] //for each field/value pair, must find each qckey in
+												 //the map object mymap
+				qcval := pair[1]
+				for okey, ovalue := range mymap {
+					if qckey == okey && qcval == ovalue {
+						jointQueryResults[i] = true
+					}
+				}
+				i += 1
+			}
+			allTrue := true
+			for _, b := range jointQueryResults {
+				if !b {
+					allTrue = false
+				}
+			}
+			if allTrue{
```

Comment:
This can be simplified to just: return allTrue
instead of:

if allTrue {
    return true
}

Reply:
Doing return true actually breaks the bisect method because I use return to break out of the loop early, when I found an element in mymap (the slice of map[string]string), that satisfied the attributeCombinedQuery . So if I never find an element in mymap that satisfies the attributeCombinedQuery (username = "soandso" && password="foofoo"), then the loop stops and I return false 
I added this comment to the code this is a little unclear :
			// Note: I cannot just return allTrue because this breaks the outer loop ..
			// The logic goes: if allTrue is never set to True, out of all the elements in outer loop mymap,
			// then the loop will finish and return false. return allTrue with no if statement, would stop the
			// loop and only check one mymap! Need to scan through all before returning false